third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2864,12 +2963,36 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
         {
             let exp_id = exp.node_id();
             let exp_ty = self.env().get_node_type(exp_id);
-            if exp_ty.is_mutable_reference() && param_ty.is_immutable_reference() {
-                // Insert Freeze operation
-                let new_id =
-                    self.new_node_id_with_type_loc(&param_ty, &self.env().get_node_loc(exp_id));
-                *exp = ExpData::Call(new_id, Operation::Freeze, vec![exp.clone()]).into_exp();
+            self.try_freeze(&param_ty, &exp_ty, exp);
+        }
+    }
+
+    /// Inserts the freeze operation when `expected_ty` is immutable ref and ty is mutable ref
+    fn try_freeze(&self, expected_ty: &Type, ty: &Type, exp: &mut Exp) {
+        let exp_id = exp.node_id();
+        if expected_ty.is_immutable_reference() && ty.is_mutable_reference() {
+            // Insert Freeze operation
+            let new_id =
+                self.new_node_id_with_type_loc(expected_ty, &self.env().get_node_loc(exp_id));
+            *exp = ExpData::Call(new_id, Operation::Freeze, vec![exp.clone()]).into_exp();
+        }
+    }
+
+    /// Inserts the freeze operation when `exp` is a tuple expression
```

Comment:
Similar here, try to use a more functional style.
```
fn freeze_tuple_exp(&self, lhs_tys: &Vec<Type>, rhs_tys: &Vec<Type>, exp: Exp, loc: &Loc) -> Exp {
    if lhs_tys.len() != rhs_tys.len() || lhs_tys == rhs_tys {
        return exp
    }
    let exp_id = exp.node_id();
    if let ExpData::Call(_, Operation::Tuple, rhs_vec) = exp.into() {
        let new_rhs = lhs_tys.iter().zip(rhs_tys.iter()).zip(rhs_vec)
            .map(|((lh_ty, rh_ty), rh)| { 
                self.try_freeze(lh_ty, rh_ty, rh)
            };
        let new_id_tuple = ...;
        ExpData::Call(new_id_tuple, Operation_Tuple, new_rhs).into_exp();
    } else {
        exp
    }
}
```

Reply:
It turns out that the if-else requires an explicit clone:
```
 if let ExpData::Call(_, Operation::Tuple, rhs_vec) = exp.clone().into() {
...
```
Not sure in this case should we still use this style?

Reply:
You can use probably use `exp.as_ref()` and then `rhs_vec` will be a ref and you can clone `rh` as needed.  In any case, `Exp` is interned, so cloning is fairly shallow anyway.

Reply:
done